From 848a93211be5798141700256b14f862a434d5bfa Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:38:38 -0500
Subject: [drm] i915: Wait for vblank before enabling the TV encoder

Message-id: <1418852380-13061-28-git-send-email-rclark@redhat.com>
Patchwork-id: 102172
O-Subject: [RHEL7 drm 27/89] drm/i915: Wait for vblank before enabling the TV encoder
Bugzilla: 1153301
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1153301
Upstream: since v3.16.4

commit f54bd2d7610032064327098a2da3c623db74722a

Author:     Ville Syrjala <ville.syrjala@linux.intel.com>
AuthorDate: Mon Sep 8 17:43:01 2014 +0300
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Wed Dec 10 17:15:02 2014 -0500

    drm/i915: Wait for vblank before enabling the TV encoder

    commit 7a98948f3b536ca9a077e84966ddc0e9f53726df upstream.

    The vblank waits in intel_tv_detect_type() are timing out for some
    reason. This is a regression caused removing seemingly useless vblank
    waits from the modeset seqeuence in:

     commit 56ef52cad5e37fca89638e4bad598a994ecc3d9f
     Author: Ville Syrjala <ville.syrjala@linux.intel.com>
     Date:   Thu May 8 19:23:15 2014 +0300

        drm/i915: Kill vblank waits after pipe enable on gmch platforms

    So it turns out they weren't all entirely useless. Apparently the pipe
    has to go through one full frame before we enable the TV port. Add a
    vblank wait to intel_enable_tv() to make sure that happens.

    Another approach was attempted by placing the vblank wait just after
    enabling the port. The theory behind that attempt was that we need to
    let the port stay enabled for one full frame before disabling it again
    during load detection. But that didn't work, and we definitely must
    have the vblank wait before enabling the port.

    Cc: Alan Bartlett <ajb@elrepo.org>
    Tested-by: Alan Bartlett <ajb@elrepo.org>
    Bugzilla: https://bugzilla.kernel.org/show_bug.cgi?id=79311
    Signed-off-by: Ville Syrjala <ville.syrjala@linux.intel.com>
    Reviewed-by: Daniel Vetter <daniel@ffwll.ch>
    Signed-off-by: Jani Nikula <jani.nikula@intel.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Signed-off-by: Rob Clark <rclark@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/i915/intel_tv.c b/drivers/gpu/drm/i915/intel_tv.c
index de8ccab..5c6f7e2 100644
--- a/drivers/gpu/drm/i915/intel_tv.c
+++ b/drivers/gpu/drm/i915/intel_tv.c
@@ -854,6 +854,10 @@ intel_enable_tv(struct intel_encoder *encoder)
  struct drm_device *dev = encoder->base.dev;
  struct drm_i915_private *dev_priv = dev->dev_private;
 
+ /* Prevents vblank waits from timing out in intel_tv_detect_type() */
+ intel_wait_for_vblank(encoder->base.dev,
+         to_intel_crtc(encoder->base.crtc)->pipe);
+
  I915_WRITE(TV_CTL, I915_READ(TV_CTL) | TV_ENC_ENABLE);
 }
 
-- 
1.7.1